### ISSUE TITLE
init:stop() takes 5 sec when logger has file backed handlers

### ISSUE THREAD

**Issue by shino** [erlang/otp#5780]
Expectation: init:stop() finishes reasonably quickly.

Actual: it takes 5 seconds.

Steps to reproduce:
1. git clone https://github.com/shino/slow_stop
2. cd slow_stop
3. rebar3 as dev release
4. _build/dev/rel/slow_stop/bin/slow_stop daemon
5. time _build/dev/rel/slow_stop/bin/slow_stop stop

At final step, it takes 4 or 5 seconds.

Additional note:
If the logger settings in sys.config is removed, stop takes ~2 seconds.

**Comment by shino** [comment]
Oops, missing environment information:
- otp 25.0-rc1
- arch linux 5.16.13

------

A little additional info:
After `init:stop()` in console, `Logger - error: {removed_failing_handler,default}` cames up.

```
% _build/dev/rel/slow_stop/bin/slow_stop console
Exec: /home/shino/local/otp/25.0-rc1/erts-13.0/bin/erlexec -boot /home/shino/g/slow_stop/_build/dev/rel/slow_stop/releases/0.0.0/start -mode interactive -boot_var SYSTEM_LIB_DIR /home/shino/local/otp/25.0-rc1/lib -config /home/shino/g/slow_stop/_build/dev/rel/slow_stop/releases/0.0.0/sys.config -args_file /home/shino/g/slow_stop/_build/dev/rel/slow_stop/releases/0.0.0/vm.args -- console
Root: /home/shino/g/slow_stop/_build/dev/rel/slow_stop
/home/shino/g/slow_stop/_build/dev/rel/slow_stop
Erlang/OTP 25 [RELEASE CANDIDATE 1] [erts-13.0] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit:ns]

Eshell V13.0  (abort with ^G)
(slow_stop@shino-xps)1> init:stop().
ok
(slow_stop@shino-xps)2> Logger - error: {removed_failing_handler,default}
```

**Comment by shino** [comment]
Quick update.

- With OTP 24.3.2, stop takes ~ 2.3 sec
- With OTP 25.0-rc1, stop takes ~ 4.3 sec

----

After setting `[{kernel,[{logger_level,all}]}]`, the logger output after init:stop() does not change.

```
Eshell V13.0  (abort with ^G)
(slow_stop@shino-xps)1> application:get_env(kernel, logger_level).
{ok,all}
(slow_stop@shino-xps)2> init:stop().
ok
(slow_stop@shino-xps)3> Logger - error: {removed_failing_handler,default}
```

**Comment by shino** [comment]
Tried to debug at some point.
This may be a race around timer server (?)

Used redbug to tracing: https://hexdocs.pm/redbug/
init:stop() with tracing `timer` module function calls, 25.0-rc1, is as follows.

```
(default) % _build/dev/rel/slow_stop/bin/slow_stop console
Exec: /home/shino/local/otp/25.0-rc1/erts-13.0/bin/erlexec -boot /home/shino/g/slow_stop/_build/dev/rel/slow_stop/releases/0.0.0/start -mode interactive -boot_var SYSTEM_LIB_DIR /home/shino/local/otp/25.0-rc1/lib -config /home/shino/g/slow_stop/_build/dev/rel/slow_stop/releases/0.0.0/sys.config -args_file /home/shino/g/slow_stop/_build/dev/rel/slow_stop/releases/0.0.0/vm.args -- console
Root: /home/shino/g/slow_stop/_build/dev/rel/slow_stop
/home/shino/g/slow_stop/_build/dev/rel/slow_stop
Erlang/OTP 25 [RELEASE CANDIDATE 1] [erts-13.0] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit:ns]

Eshell V13.0  (abort with ^G)
(slow_stop@127.0.0.1)1>
(slow_stop@127.0.0.1)1>
(slow_stop@127.0.0.1)1> redbug:start("timer:_ -> return,stack", [{max_msg_size, 1000000}, {print_depth, 30}, {msgs, 100}, {print_msec, true}, {time, 100*1000}]).
{59,30}
(slow_stop@127.0.0.1)2> init:stop().
ok

% 15:06:39.067 <0.87.0>(dead)
% timer:terminate(shutdown, #Ref<0.4049152735.3242328065.17831>)
%   gen_server:try_terminate/3
%   gen_server:terminate/10
%   proc_lib:init_p_do_apply/3

% 15:06:39.067 <0.87.0>(dead)
% timer:terminate/2 -> ok

% 15:06:39.067 <0.81.0>(logger_std_h_default)
% timer:cancel({once,#Ref<0.4049152735.3242196993.17834>})
%   logger_h_common:cancel_repeated_filesync/1
%   logger_h_common:do_terminate/2
%   logger_olp:try_callback_call/4
%   logger_olp:terminate/2
%   gen_server:try_terminate/3
%   gen_server:terminate/10
%   proc_lib:init_p_do_apply/3

% 15:06:41.068 <0.81.0>(dead)
% timer:cancel/1 -> {exit,killed}
(slow_stop@127.0.0.1)3> Logger - error: {removed_failing_handler,default}
% 15:06:42.071 <0.102.0>(user)
% timer:sleep(50)
%   group_history:wait_for_kernel_safe_sup/0
%   group_history:load/0
%   group:server/3
(slow_stop@127.0.0.1)3>
(default) %
```

- timer server terminated after init:stop()
- Just after that, timer:cancel() is called (https://github.com/erlang/otp/blob/master/lib/kernel/src/logger_h_common.erl#L457)
- After 2 seconds, timer:cancel() exited with {exit, killed}

------

With 24.3.2, timer:cancel() comes first and finish immediately, then timer:terminate() is called.
I don't yet find why such difference.

```
(default) % _build/dev/rel/slow_stop/bin/slow_stop console
Exec: /home/shino/local/otp/24.3.2/erts-12.3/bin/erlexec -boot /home/shino/g/slow_stop_24/_build/dev/rel/slow_stop/releases/0.0.0/start -mode interactive -boot_var SYSTEM_LIB_DIR /home/shino/local/otp/24.3.2/lib -config /home/shino/g/slow_stop_24/_build/dev/rel/slow_stop/releases/0.0.0/sys.config -args_file /home/shino/g/slow_stop_24/_build/dev/rel/slow_stop/releases/0.0.0/vm.args -- console
Root: /home/shino/g/slow_stop_24/_build/dev/rel/slow_stop
/home/shino/g/slow_stop_24/_build/dev/rel/slow_stop
Erlang/OTP 24 [erts-12.3] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [jit]

Eshell V12.3  (abort with ^G)
(search)`redbu': redbug:start("timer:_ -> return,stack", [{max_msg_size, 1000000}, {print_depth, 30}, {msgs, 100}, {print_msec, true}, {time, 100*1000}]).
{59,31}
(slow_stop@shino-xps)2> init:stop().
ok

% 15:13:49.542 <0.80.0>(logger_std_h_default)
% timer:cancel({-576460746770441,#Ref<0.1495740804.2437939202.127907>})
%   logger_h_common:cancel_repeated_filesync/1
%   logger_h_common:do_terminate/2
%   logger_olp:try_callback_call/4
%   logger_olp:terminate/2
%   gen_server:try_terminate/3
%   gen_server:terminate/10
%   proc_lib:init_p_do_apply/3

% 15:13:49.542 <0.87.0>(timer_server)
% timer:handle_call({cancel,{-576460746770441,#Ref<0.1495740804.2437939202.127907>},
        -576460747412478}, {<0.80.0>,#Ref<0.1495740804.2437939202.127952>}, [])
%   gen_server:try_handle_call/4
%   gen_server:handle_msg/6
%   proc_lib:init_p_do_apply/3

% 15:13:49.542 <0.87.0>(timer_server)
% timer:handle_call/3 -> {reply,{ok,cancel},[],infinity}

% 15:13:49.542 <0.80.0>(logger_std_h_default)
% timer:cancel/1 -> {ok,cancel}

% 15:13:49.548 <0.87.0>(dead)
% timer:terminate(shutdown, [])
%   gen_server:try_terminate/3
%   gen_server:terminate/10
%   proc_lib:init_p_do_apply/3

% 15:13:49.548 <0.87.0>(dead)
% timer:terminate/2 -> ok

% 15:13:50.549 <0.100.0>(user)
% timer:sleep(50)
%   group_history:wait_for_kernel_safe_sup/0
%   group_history:load/0
%   group:server/3
(slow_stop@shino-xps)3> %                                                                                                                                                                                         ```

**Comment by shino** [comment]
erlang:send_after may be better than timer:apply_after here??

**Comment by garazdawi** [comment]
Great find! #4811 changed the placement of the timer server in the kernel supervisor, which means that it will terminate before logger instead of after as it did before.

Changing `logger_h_common.erl` to use erlang:send_after would work. I assume it was not done originally because any message we would send using send_after would end up in the handle_info clause, which is dedicated to forwarding messages to the callback module. We should be able to work around that by saving a ref in the logger handler state to match on for the file sync.

Do you want to take a stab at implementing it, or should I add it to my todo list?

**Comment by shino** [comment]
Thanks for response!

> Do you want to take a stab at implementing it, or should I add it to my todo list?

I'm happy if it's possible for you to add this issue to your list.
(I'm now hunting other bugs 😅 )

**Comment by garazdawi** [comment]
Fix is available in #5829, please test it to make sure it solves your problem!

**Comment by shino** [comment]
With your branch, stop returns quickly (~ 2 sec) :smile: :+1: 
Thanks for your support and fix!!
